File: MATLAB/examples/json_utils_example.m
%% JSON Utilities Example
%  Demonstrates the usage of JSON parsing and generation utilities
%  for MATLAB.
%
%  Run this script in MATLAB to see examples of:
%  - Parsing JSON strings
%  - Generating JSON from MATLAB data
%  - Working with nested structures
%  - File I/O operations
%  - Utility functions
%
%  Usage:
%      run('json_utils_example.m')

fprintf('========================================\n');
fprintf('  JSON Utilities Examples\n');
fprintf('========================================\n\n');

%% Initialize
utils = json_utils.mod();

%% Example 1: Parse a simple JSON object
fprintf('Example 1: Parse Simple JSON Object\n');
fprintf('------------------------------------\n');

jsonStr = '{"name": "John Doe", "age": 30, "city": "New York"}';
fprintf('Input: %s\n\n', jsonStr);

data = utils.parse(jsonStr);
fprintf('Parsed data:\n');
fprintf('  name: %s\n', data.name);
fprintf('  age: %d\n', data.age);
fprintf('  city: %s\n\n', data.city);

%% Example 2: Parse a nested JSON object
fprintf('Example 2: Parse Nested JSON Object\n');
fprintf('------------------------------------\n');

jsonStr = ['{\n', ...
           '  "user": {\n', ...
           '    "id": 12345,\n', ...
           '    "profile": {\n', ...
           '      "name": "Alice",\n', ...
           '      "email": "alice@example.com"\n', ...
           '    },\n', ...
           '    "roles": ["admin", "editor"]\n', ...
           '  }\n', ...
           '}'];
fprintf('Input:\n%s\n\n', jsonStr);

data = utils.parse(jsonStr);
fprintf('Accessing nested data:\n');
fprintf('  user.id: %d\n', data.user.id);
fprintf('  user.profile.name: %s\n', data.user.profile.name);
fprintf('  user.profile.email: %s\n', data.user.profile.email);
fprintf('  user.roles{1}: %s\n\n', data.user.roles{1});

%% Example 3: Parse JSON arrays
fprintf('Example 3: Parse JSON Arrays\n');
fprintf('------------------------------------\n');

% Array of numbers
jsonStr = '[1, 2, 3, 4, 5]';
numbers = utils.parse(jsonStr);
fprintf('Numbers array: ');
for i = 1:length(numbers)
    fprintf('%d ', numbers{i});
end
fprintf('\n');

% Array of strings
jsonStr = '["apple", "banana", "cherry"]';
fruits = utils.parse(jsonStr);
fprintf('Fruits array: ');
for i = 1:length(fruits)
    fprintf('%s ', fruits{i});
end
fprintf('\n\n');

%% Example 4: Parse special values
fprintf('Example 4: Parse Special Values\n');
fprintf('------------------------------------\n');

jsonStr = '{"active": true, "deleted": false, "data": null, "count": 0}';
data = utils.parse(jsonStr);
fprintf('active (boolean): %d\n', data.active);
fprintf('deleted (boolean): %d\n', data.deleted);
fprintf('data (null): %s\n', mat2str(data.data));
fprintf('count (zero): %d\n\n', data.count);

%% Example 5: Encode MATLAB struct to JSON
fprintf('Example 5: Encode Struct to JSON\n');
fprintf('------------------------------------\n');

person = struct();
person.name = 'Bob';
person.age = 25;
person.hobbies = {{'reading', 'gaming', 'coding'}};

json = utils.encode(person);
fprintf('Compact JSON:\n%s\n\n', json);

prettyJson = utils.encodePretty(person);
fprintf('Pretty JSON:\n%s\n\n', prettyJson);

%% Example 6: Encode cell array to JSON
fprintf('Example 6: Encode Cell Array to JSON\n');
fprintf('------------------------------------\n');

mixedArray = {42, 'hello', true, struct('x', 1, 'y', 2)};
json = utils.encode(mixedArray);
fprintf('Mixed array as JSON:\n%s\n\n', json);

%% Example 7: Round-trip (parse then encode)
fprintf('Example 7: Round-trip Test\n');
fprintf('------------------------------------\n');

original = struct();
original.product = 'Laptop';
original.price = 999.99;
original.inStock = true;
original.specs = struct();
original.specs.cpu = 'Intel i7';
original.specs.ram = '16GB';

json = utils.encode(original);
fprintf('Original encoded:\n%s\n\n', json);

restored = utils.parse(json);
fprintf('Restored data:\n');
fprintf('  product: %s\n', restored.product);
fprintf('  price: %.2f\n', restored.price);
fprintf('  inStock: %d\n', restored.inStock);
fprintf('  specs.cpu: %s\n', restored.specs.cpu);
fprintf('  specs.ram: %s\n\n', restored.specs.ram);

%% Example 8: Using get with default values
fprintf('Example 8: Safe Access with Defaults\n');
fprintf('------------------------------------\n');

config = utils.parse('{"host": "localhost", "port": 8080}');

% Existing key
host = utils.get(config, 'host', '127.0.0.1');
fprintf('host (exists): %s\n', host);

% Missing key with default
timeout = utils.get(config, 'timeout', 30);
fprintf('timeout (default): %d\n\n', timeout);

%% Example 9: Using getPath for nested access
fprintf('Example 9: Nested Path Access\n');
fprintf('------------------------------------\n');

jsonStr = ['{\n', ...
           '  "database": {\n', ...
           '    "connection": {\n', ...
           '      "host": "db.example.com",\n', ...
           '      "port": 5432\n', ...
           '    }\n', ...
           '  }\n', ...
           '}'];

data = utils.parse(jsonStr);

% Access nested value
host = utils.getPath(data, 'database.connection.host', 'localhost');
port = utils.getPath(data, 'database.connection.port', 3306);
missing = utils.getPath(data, 'database.connection.ssl', false);

fprintf('database.connection.host: %s\n', host);
fprintf('database.connection.port: %d\n', port);
fprintf('database.connection.ssl (default): %d\n\n', missing);

%% Example 10: Validation
fprintf('Example 10: JSON Validation\n');
fprintf('------------------------------------\n');

validJson = '{"valid": true}';
invalidJson = 'this is not json';

fprintf('Is ''%s'' valid? %d\n', validJson, utils.isValid(validJson));
fprintf('Is ''%s'' valid? %d\n\n', invalidJson, utils.isValid(invalidJson));

%% Example 11: Safe parsing with parseOrNull
fprintf('Example 11: Safe Parsing\n');
fprintf('------------------------------------\n');

% This will succeed
data = utils.parseOrNull('{"success": true}');
if ~isempty(data)
    fprintf('Parsed successfully: success=%d\n', data.success);
end

% This will return empty without throwing
data = utils.parseOrNull('not valid json');
if isempty(data)
    fprintf('Invalid JSON handled gracefully\n');
end
fprintf('\n');

%% Example 12: Minify JSON
fprintf('Example 12: Minify JSON\n');
fprintf('------------------------------------\n');

prettyJson = ['{\n', ...
              '  "name": "Test",\n', ...
              '  "items": [\n', ...
              '    1,\n', ...
              '    2,\n', ...
              '    3\n', ...
              '  ]\n', ...
              '}'];

minified = utils.minify(prettyJson);
fprintf('Before:\n%s\n\n', prettyJson);
fprintf('After minification:\n%s\n\n', minified);

%% Example 13: Static methods
fprintf('Example 13: Static Quick Methods\n');
fprintf('------------------------------------\n');

% Quick parse without creating instance
data = json_utils.mod.quickParse('{"quick": "parse"}');
fprintf('quickParse result: %s\n', data.quick);

% Quick encode without creating instance
json = json_utils.mod.quickEncode(struct('quick', 'encode'), true);
fprintf('quickEncode result:\n%s\n\n', json);

%% Summary
fprintf('========================================\n');
fprintf('  Examples completed!\n');
fprintf('========================================\n');

File: PHP/examples/json_utils_example.php
<?php
/**
 * AllToolkit - JSON Utilities Example
 *
 * Demonstrates various use cases for JsonUtils
 *
 * @author AllToolkit Contributors
 * @license MIT
 */

require_once __DIR__ . '/../json_utils/mod.php';

use AllToolkit\JsonUtils;

echo "AllToolkit - JSON Utilities Examples\n";
echo "====================================\n\n";

// Example 1: Basic parsing and stringification
echo "1. Basic Parsing and Stringification\n";
echo "------------------------------------\n";

$jsonString = '{"name":"John","age":30,"city":"New York"}';
$data = JsonUtils::parse($jsonString);

echo "Input JSON: $jsonString\n";
echo "Parsed name: " . $data['name'] . "\n";
echo "Parsed age: " . $data['age'] . "\n\n";

// Convert back to JSON
$backToJson = JsonUtils::stringify($data);
echo "Stringified: $backToJson\n\n";

// Example 2: Pretty printing
echo "2. Pretty Printing\n";
echo "------------------\n";

$minified = '{"users":[{"name":"Alice"},{"name":"Bob"}]}';
$pretty = JsonUtils::prettyPrint($minified);
echo "Minified:\n$minified\n\n";
echo "Pretty printed:\n$pretty\n\n";

// Example 3: Nested data access with dot notation
echo "3. Nested Data Access (Dot Notation)\n";
echo "------------------------------------\n";

$nestedData = [
    'company' => [
        'name' => 'Acme Corp',
        'address' => [
            'street' => '123 Main St',
            'city' => 'Boston',
            'zip' => '02101'
        ],
        'employees' => [
            ['name' => 'Alice', 'role' => 'Developer'],
            ['name' => 'Bob', 'role' => 'Designer']
        ]
    ]
];

echo "Company name: " . JsonUtils::get($nestedData, 'company.name') . "\n";
echo "City: " . JsonUtils::get($nestedData, 'company.address.city') . "\n";
echo "First employee: " . JsonUtils::get($nestedData, 'company.employees.0.name') . "\n";
echo "Missing path with default: " . JsonUtils::get($nestedData, 'company.phone', 'N/A') . "\n\n";

// Example 4: Setting nested values
echo "4. Setting Nested Values\n";
echo "------------------------\n";

$config = [];
JsonUtils::set($config, 'database.host', 'localhost');
JsonUtils::set($config, 'database.port', 3306);
JsonUtils::set($config, 'database.credentials.username', 'admin');
JsonUtils::set($config, 'database.credentials.password', 'secret');

echo "Config after setting values:\n";
echo JsonUtils::stringify($config, true) . "\n\n";

// Example 5: Checking if path exists
echo "5. Checking Path Existence\n";
echo "--------------------------\n";

$hasHost = JsonUtils::has($config, 'database.host');
$hasSsl = JsonUtils::has($config, 'database.ssl');

echo "Has database.host: " . ($hasHost ? 'Yes' : 'No') . "\n";
echo "Has database.ssl: " . ($hasSsl ? 'Yes' : 'No') . "\n\n";

// Example 6: Merging JSON data
echo "6. Merging JSON Data\n";
echo "--------------------\n";

$defaults = [
    'theme' => 'light',
    'language' => 'en',
    'notifications' => [
        'email' => true,
        'push' => false
    ]
];

$userSettings = [
    'theme' => 'dark',
    'notifications' => [
        'push' => true
    ]
];

$merged = JsonUtils::merge($defaults, $userSettings);
echo "Merged settings:\n";
echo JsonUtils::stringify($merged, true) . "\n\n";

// Example 7: Flatten and unflatten
echo "7. Flatten and Unflatten\n";
echo "------------------------\n";

$nested = [
    'user' => [
        'profile' => [
            'name' => 'John',
            'settings' => [
                'theme' => 'dark',
                'notifications' => true
            ]
        ]
    ]
];

$flat = JsonUtils::flatten($nested);
echo "Flattened:\n";
print_r($flat);

$unflattened = JsonUtils::unflatten($flat);
echo "\nUnflattened:\n";
echo JsonUtils::stringify($unflattened, true) . "\n\n";

// Example 8: Type-safe getters
echo "8. Type-safe Getters\n";
echo "--------------------\n";

$apiResponse = [
    'status' => 'success',
    'code' => '200',
    'data' => [
        'count' => '42',
        'average' => '85.5',
        'items' => ['a', 'b', 'c']
    ]
];

echo "Status (string): " . JsonUtils::getString($apiResponse, 'status') . "\n";
echo "Code (int): " . JsonUtils::getInt($apiResponse, 'code') . "\n";
echo "Count (int): " . JsonUtils::getInt($apiResponse, 'data.count') . "\n";
echo "Average (float): " . JsonUtils::getFloat($apiResponse, 'data.average') . "\n";
echo "Items count: " . count(JsonUtils::getArray($apiResponse, 'data.items')) . "\n\n";

// Example 9: Validation
echo "9. JSON Validation\n";
echo "------------------\n";

$validJson = '{"name":"test","value":123}';
$invalidJson = '{"name":"test",}';

echo "Is valid JSON: " . (JsonUtils::isValid($validJson) ? 'Yes' : 'No') . "\n";
echo "Is invalid JSON: " . (JsonUtils::isValid($invalidJson) ? 'Yes' : 'No') . "\n";

$error = JsonUtils::getError($invalidJson);
echo "Error message: $error\n\n";

// Example 10: Working with arrays
echo "10. Working with Arrays\n";
echo "-----------------------\n";

$users = [
    ['id' => 1, 'name' => 'Alice', 'active' => true],
    ['id' => 2, 'name' => 'Bob', 'active' => false],
    ['id' => 3, 'name' => 'Charlie', 'active' => true]
];

$keys = JsonUtils::keys($users[0]);
echo "User object keys: " . implode(', ', $keys) . "\n";

$names = array_map(function($user) {
    return $user['name'];
}, $users);
echo "All names: " . implode(', ', $names) . "\n\n";

// Example 11: Minification
echo "11. Minification\n";
echo "----------------\n";

$prettyJson = JsonUtils::stringify(['a' => 1, 'b' => 2], true);
echo "Pretty:\n$prettyJson\n";

$minified = JsonUtils::minify($prettyJson);
echo "\nMinified: $minified\n\n";

// Example 12: Cloning
echo "12. Deep Cloning\n";
echo "----------------\n";

$original = [
    'user' => [
        'name' => 'John',
        'preferences' => ['theme' => 'dark']
    ]
];

$clone = JsonUtils::clone($original);
$clone['user']['name'] = 'Jane';
$clone['user']['preferences']['theme'] = 'light';

echo "Original name: " . $original['user']['name'] . "\n";
echo "Clone name: " . $clone['user']['name'] . "\n";
echo "Original theme: " . $original['user']['preferences']['theme'] . "\n";
echo "Clone theme: " . $clone['user']['preferences']['theme'] . "\n\n";

// Example 13: Comparing JSON
echo "13. Comparing JSON\n";
echo "------------------\n";

$obj1 = ['a' => 1, 'b' => ['c' => 2]];
$obj2 = ['b' => ['c' => 2], 'a' => 1];
$obj3 = ['a' => 1, 'b' => ['c' => 3]];

echo "obj1 equals obj2: " . (JsonUtils::equals($obj1, $obj2) ? 'Yes' : 'No') . "\n";
echo "obj1 equals obj3: " . (JsonUtils::equals($obj1, $obj3) ? 'Yes' : 'No') . "\n\n";

echo "====================================\n";
echo "Examples completed!\n";
